Add splinter

oriram opened the pull request:
# What does this PR do?

[Splinter](https://arxiv.org/abs/2101.00438) implementation
@patil-suraj @LysandreJik @patrickvonplaten

yuvalkirstain committed:
splinter template

yuvalkirstain committed:
initialize splinter classes

oriram committed:
Splinter Tokenizer

oriram committed:
splinter.rst

yuvalkirstain committed:
tokenization fixes

yuvalkirstain committed:
Merge remote-tracking branch 'origin/master'

oriram committed:
Documentation & some minor variable name changes

oriram committed:
bug fix (added back question_token_id to config) + variable names

oriram committed:
Minor bug fixes + variable name changes

yuvalkirstain committed:
update transformers

oriram committed:
Fix Splinter references after merge with new transformers

oriram committed:
changes after running make style & quality

oriram committed:
Fix documentation unindent

oriram committed:
Fix doc indentation in tokenization_splinter

oriram committed:
Fix also SplinterTokenizerFast

oriram committed:
Add Splinter to index.rst and README

oriram committed:
Fixdouble whitespace from index.rst

oriram committed:
Fixed index.rst with 'make fix-copies'

patil-suraj reviewed (commented):
Thanks a lot for the PR! It's looking good.

I just left a few comments, specifically 

- As most of the code is a copy of the BERT model we should add `Copied from XXX` statements. I've left the details in the comment.
- Also `Splinter` is specific for QA,  no? Does it support other tasks like sequence classification, multiple-choice, etc? If not then maybe we could remove those heads. wdyt @LysandreJik @sgugger 

Also to upload the models you could create an organization on the hub https://huggingface.co/organizations/new. And then we could use the `push_to_hub` API to push the model checkpoints on the hub, here are the [docs](https://huggingface.co/transformers/master/model_sharing.html#push-your-model-from-python)

sgugger commented on README.md:
```diff
@@ -262,6 +262,7 @@ Min, Patrick Lewis, Ledell Wu, Sergey Edunov, Danqi Chen, and Wen-tau Yih.
 1. **[RoBERTa](https://huggingface.co/transformers/model_doc/roberta.html)** (from Facebook), released together with the paper a [Robustly Optimized BERT Pretraining Approach](https://arxiv.org/abs/1907.11692) by Yinhan Liu, Myle Ott, Naman Goyal, Jingfei Du, Mandar Joshi, Danqi Chen, Omer Levy, Mike Lewis, Luke Zettlemoyer, Veselin Stoyanov.
 1. **[RoFormer](https://huggingface.co/transformers/model_doc/roformer.html)** (from ZhuiyiTechnology), released together with the paper a [RoFormer: Enhanced Transformer with Rotary Position Embedding](https://arxiv.org/pdf/2104.09864v1.pdf) by Jianlin Su and Yu Lu and Shengfeng Pan and Bo Wen and Yunfeng Liu.
 1. **[SpeechToTextTransformer](https://huggingface.co/transformers/model_doc/speech_to_text.html)** (from Facebook), released together with the paper [fairseq S2T: Fast Speech-to-Text Modeling with fairseq](https://arxiv.org/abs/2010.05171) by Changhan Wang, Yun Tang, Xutai Ma, Anne Wu, Dmytro Okhonko, Juan Pino.
+1. **[Splinter](https://huggingface.co/transformers/model_doc/splinter.html)** (from Tel Aviv University), released together with the paper [Few-Shot Question Answering by Pretraining Span Selection](https://arxiv.org/abs/2101.00438) by Ori Ram, Yuval Kirstain, Jonathan Berant, Amir Globerson, Omer Levy.
```
```suggestion
1. **[Splinter](https://huggingface.co/transformers/model_doc/master/splinter.html)** (from Tel Aviv University), released together with the paper [Few-Shot Question Answering by Pretraining Span Selection](https://arxiv.org/abs/2101.00438) by Ori Ram, Yuval Kirstain, Jonathan Berant, Amir Globerson, Omer Levy.
```

The model doc won't be in the stable doc once this is merged.

sgugger commented on src/transformers/models/splinter/configuration_splinter.py:
```diff
@@ -0,0 +1,133 @@
+logger = logging.get_logger(__name__)
+
+SPLINTER_PRETRAINED_CONFIG_ARCHIVE_MAP = {
+    "splinter-base": "https://huggingface.co/splinter-base/resolve/main/config.json",
+    "splinter-large": "https://huggingface.co/splinter-large/resolve/main/config.json",
```
Those URLs do not work

oriram replied:
@sgugger can "tau/splinter-base" and "tau/splinter-large" work? 
I created an organization for Tel Aviv University at the model hub under "tau".

sgugger replied:
Yes, you will just need to upload the model there :-)

sgugger commented on src/transformers/models/splinter/modeling_splinter.py:
```diff
@@ -0,0 +1,1629 @@
+_TOKENIZER_FOR_DOC = "SplinterTokenizer"
+
+SPLINTER_PRETRAINED_MODEL_ARCHIVE_LIST = [
+    "splinter-base",
+    "splinter-large",
```
Those are not valid model identifiers.

sgugger commented on src/transformers/models/splinter/modeling_splinter.py:
```diff
@@ -0,0 +1,1629 @@
+]
+
+
+def load_tf_weights_in_splinter(model, config, tf_checkpoint_path):
+    """Load tf checkpoints in a pytorch model."""
+    try:
+        import re
+
+        import numpy as np
+        import tensorflow as tf
```
Yes, the whole function should be removed.

sgugger commented on src/transformers/models/splinter/modeling_splinter.py:
```diff
@@ -0,0 +1,1629 @@
+
+SPLINTER_INPUTS_DOCSTRING = r"""
+    Args:
+        input_ids (:obj:`torch.LongTensor` of shape :obj:`{0}`):
```
Oh there is a bug in the model template, there should be parenthesis here:
```suggestion
        input_ids (:obj:`torch.LongTensor` of shape :obj:`({0})`):
```
and below at each `{0}` (see the BERT modeling file for an example).

sgugger commented on src/transformers/models/splinter/modeling_splinter.py:
```diff
@@ -0,0 +1,1629 @@
+            - 1 indicates the head is **not masked**,
+            - 0 indicates the head is **masked**.
+
+        inputs_embeds (:obj:`torch.FloatTensor` of shape :obj:`(batch_size, sequence_length, hidden_size)`, `optional`):
```
```suggestion
        inputs_embeds (:obj:`torch.FloatTensor` of shape :obj:`({0}, hidden_size)`, `optional`):
```
That's the reason why there should be parenthesis on all {0} above.

sgugger commented on src/transformers/models/splinter/modeling_splinter.py:
```diff
@@ -0,0 +1,1629 @@
+
+    def _prune_heads(self, heads_to_prune):
+        """
+        Prunes heads of the model. heads_to_prune: dict of {layer_num: list of heads to prune in this layer} See base
+        class PreTrainedModel
```
This needs to be properly formatted.

oriram replied:
Thanks @sgugger. What do you mean by properly formatted? This was auto-generated from the template :)

sgugger replied:
The `heads_to_prune` should be on a new line probably.

sgugger commented on src/transformers/models/splinter/modeling_splinter.py:
```diff
@@ -0,0 +1,1629 @@
+        for layer, heads in heads_to_prune.items():
+            self.encoder.layer[layer].attention.prune_heads(heads)
+
+    @add_start_docstrings_to_model_forward(SPLINTER_INPUTS_DOCSTRING.format("(batch_size, sequence_length)"))
```
```suggestion
    @add_start_docstrings_to_model_forward(SPLINTER_INPUTS_DOCSTRING.format("batch_size, sequence_length"))
```

sgugger commented on src/transformers/models/splinter/modeling_splinter.py:
```diff
@@ -0,0 +1,1629 @@
+    def set_output_embeddings(self, new_embeddings):
+        self.cls.predictions.decoder = new_embeddings
+
+    @add_start_docstrings_to_model_forward(SPLINTER_INPUTS_DOCSTRING.format("(batch_size, sequence_length)"))
```
```suggestion
    @add_start_docstrings_to_model_forward(SPLINTER_INPUTS_DOCSTRING.format("batch_size, sequence_length"))
```

sgugger commented on src/transformers/models/splinter/modeling_splinter.py:
```diff
@@ -0,0 +1,1629 @@
+
+        self.init_weights()
+
+    @add_start_docstrings_to_model_forward(SPLINTER_INPUTS_DOCSTRING.format("(batch_size, sequence_length)"))
```
```suggestion
    @add_start_docstrings_to_model_forward(SPLINTER_INPUTS_DOCSTRING.format("batch_size, sequence_length"))
```

sgugger commented on src/transformers/models/splinter/modeling_splinter.py:
```diff
@@ -0,0 +1,1629 @@
+        """Controls whether we use the pretrained QASS layer's parameters, or randomly initialized ones"""
+        return self.splinter_qass if not self.initialize_new_qass else self.new_splinter_qass
+
+    @add_start_docstrings_to_model_forward(SPLINTER_INPUTS_DOCSTRING.format("(batch_size, sequence_length)"))
```
```suggestion
    @add_start_docstrings_to_model_forward(SPLINTER_INPUTS_DOCSTRING.format("batch_size, sequence_length"))
```

sgugger commented on src/transformers/models/splinter/tokenization_splinter.py:
```diff
@@ -0,0 +1,164 @@
+
+PRETRAINED_VOCAB_FILES_MAP = {
+    "vocab_file": {
+        "splinter-base": "https://huggingface.co/splinter-base/resolve/main/vocab.txt",
+        "splinter-large": "https://huggingface.co/splinter-large/resolve/main/vocab.txt",
```
Those URLs need to be fixed.

sgugger commented on src/transformers/models/splinter/tokenization_splinter.py:
```diff
@@ -0,0 +1,164 @@
+    r"""
+    Construct a Splinter tokenizer.
+
+    :class:`~transformers.SplinterTokenizer` is identical to :class:`~transformers.BertTokenizer` and runs end-to-end
+    tokenization: punctuation splitting and wordpiece.
```
It's not exactly identical though, the difference should be mentioned here.

sgugger commented on src/transformers/models/splinter/tokenization_splinter.py:
```diff
@@ -0,0 +1,164 @@
+
+        if self.padding_side == "right":
+            # Input is question-then-context
+            return [0] + [0] * len(token_ids_0) + [0] + [0] + [0] + [1] * len(token_ids_1) + [1]
```
```suggestion
            return [0] + [0] * len(token_ids_0) + [0, 0, 0] + [1] * len(token_ids_1) + [1]
```

sgugger commented on src/transformers/models/splinter/tokenization_splinter.py:
```diff
@@ -0,0 +1,164 @@
+            return [0] + [0] * len(token_ids_0) + [0] + [0] + [0] + [1] * len(token_ids_1) + [1]
+        else:
+            # Input is context-then-question
+            return [0] + [0] * len(token_ids_0) + [0] + [1] * len(token_ids_1) + [1] + [1] + [1]
```
```suggestion
            return [0] + [0] * len(token_ids_0) + [0] + [1] * len(token_ids_1) + [1, 1, 1]
```

sgugger commented on src/transformers/models/splinter/tokenization_splinter.py:
```diff
@@ -0,0 +1,164 @@
+
+        Returns :obj:`None` if the token has not been set.
+        """
+        if self._cls_token is None:
```
Why check for the private cls token?

oriram replied:
I followed the conventions from ``tokenization_utils_base``.
I'll remove it.

sgugger commented on src/transformers/models/splinter/tokenization_splinter_fast.py:
```diff
@@ -0,0 +1,136 @@
+
+PRETRAINED_VOCAB_FILES_MAP = {
+    "vocab_file": {
+        "splinter-base": "https://huggingface.co/splinter-base/resolve/main/vocab.txt",
+        "splinter-large": "https://huggingface.co/splinter-large/resolve/main/vocab.txt",
```
These URLs need fixing as well.

oriram replied:
I wasn't sure what the URLs will be as the models haven't been uploaded yet. 
Any thoughts/suggestions? @sgugger

sgugger replied:
If you do upload them to `tau/splinter-base` and `tau/splinter-large`, you will just need to add the organization in the URLs.

sgugger commented on src/transformers/models/splinter/tokenization_splinter_fast.py:
```diff
@@ -0,0 +1,136 @@
+    r"""
+    Construct a fast Splinter tokenizer.
+
+    :class:`~transformers.SplinterTokenizer` is identical to :class:`~transformers.BertTokenizer` and runs end-to-end
```
```suggestion
    :class:`~transformers.SplinterTokenizerFast` is identical to :class:`~transformers.BertTokenizerFast` and runs end-to-end
```

Plus same comment as above: the difference with the BERT tokenizer should be highlighted.

sgugger commented on src/transformers/models/splinter/tokenization_splinter_fast.py:
```diff
@@ -0,0 +1,136 @@
+    :class:`~transformers.SplinterTokenizer` is identical to :class:`~transformers.BertTokenizer` and runs end-to-end
+    tokenization: punctuation splitting and wordpiece.
+
+    Refer to superclass :class:`~transformers.BertTokenizer` for usage examples and documentation concerning
```
```suggestion
    Refer to superclass :class:`~transformers.BertTokenizerFast` for usage examples and documentation concerning
```

sgugger commented on src/transformers/models/splinter/tokenization_splinter_fast.py:
```diff
@@ -0,0 +1,136 @@
+
+        Returns :obj:`None` if the token has not been set.
+        """
+        if self._cls_token is None:
+            return None
```
Same here, why is this checked?

sgugger reviewed (commented):
Thanks a lot for adding this model. I agree we shouldn't necessarily add all classes if the model is mostly for question answering.
In particular, the sequence classification model could be used to classify pairs of sentences with that new [QUESTION] token added there. Does it make sense and give decent results? If not, the model for sequence classification should probably be removed.

oriram committed:
Update docs/source/model_doc/splinter.rst

Co-authored-by: Suraj Patil <surajp815@gmail.com>

oriram committed:
Update docs/source/model_doc/splinter.rst

Co-authored-by: Suraj Patil <surajp815@gmail.com>

oriram committed:
Update docs/source/model_doc/splinter.rst

Co-authored-by: Suraj Patil <surajp815@gmail.com>

oriram committed:
Update docs/source/model_doc/splinter.rst

Co-authored-by: Suraj Patil <surajp815@gmail.com>

oriram committed:
Update src/transformers/models/splinter/__init__.py

Co-authored-by: Suraj Patil <surajp815@gmail.com>

oriram committed:
Added "copied from BERT" comments

oriram committed:
Removing unnexessary code from modeling_splinter

oriram committed:
Update README.md

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

oriram committed:
Update src/transformers/models/splinter/configuration_splinter.py

Co-authored-by: Suraj Patil <surajp815@gmail.com>

oriram committed:
Remove references to TF modeling from splinter

oriram committed:
Update src/transformers/models/splinter/modeling_splinter.py

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

oriram committed:
Remove unnecessary check

oriram committed:
Merge branch 'add_splinter' of https://github.com/yuvalkirstain/transformers into add_splinter

oriram committed:
Update src/transformers/models/splinter/modeling_splinter.py

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

oriram committed:
Add differences between Splinter and Bert tokenizers

oriram committed:
Merge branch 'add_splinter' of https://github.com/yuvalkirstain/transformers into add_splinter

oriram committed:
Update src/transformers/models/splinter/modeling_splinter.py

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

oriram committed:
Update src/transformers/models/splinter/tokenization_splinter_fast.py

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

oriram committed:
Remove unnecessary check

oriram committed:
Merge remote-tracking branch 'origin/add_splinter' into add_splinter

# Conflicts:
#	src/transformers/models/splinter/tokenization_splinter_fast.py

oriram committed:
Doc formatting

oriram committed:
Update src/transformers/models/splinter/tokenization_splinter.py

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

oriram committed:
Update src/transformers/models/splinter/tokenization_splinter.py

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

oriram committed:
bug fix: remove load_tf_weights attribute

patrickvonplaten commented on docs/source/model_doc/splinter.rst:
```diff
@@ -0,0 +1,120 @@
+
+Tips:
+
+- Splinter was trained to predict answers spans conditioned on a special [QUESTION] token. These tokens contextualize
```
```suggestion
- Splinter was trained to predict answer spans conditioned on a special [QUESTION] token. These tokens contextualize
```

patrickvonplaten commented on docs/source/model_doc/splinter.rst:
```diff
@@ -0,0 +1,120 @@
+  to question representations which are used to predict the answers. This layer is called QASS, and is the default
+  behaviour in the :class:`~transformers.SplinterForQuestionAnswering` class. Therefore:
+- Use :class:`~transformers.SplinterTokenizer` (rather than :class:`~transformers.BertTokenizer`), as it already contains this special token. Also, its default
+  behavior is to use this token when two sequences are given (for example, in the run_qa.py script).
```
```suggestion
  behavior is to use this token when two sequences are given (for example, in the `run_qa.py` script).
```

patrickvonplaten commented on docs/source/model_doc/splinter.rst:
```diff
@@ -0,0 +1,120 @@
+  behaviour in the :class:`~transformers.SplinterForQuestionAnswering` class. Therefore:
+- Use :class:`~transformers.SplinterTokenizer` (rather than :class:`~transformers.BertTokenizer`), as it already contains this special token. Also, its default
+  behavior is to use this token when two sequences are given (for example, in the run_qa.py script).
+- If you plan on using Splinter outside run_qa.py, please keep in mind the question token - it might be important for
```
```suggestion
- If you plan on using Splinter outside `run_qa.py`, please keep in mind the question token - it might be important for
```

patrickvonplaten commented on src/transformers/models/splinter/configuration_splinter.py:
```diff
@@ -0,0 +1,131 @@
+        initialize_new_qass (:obj:`bool`, `optional`, defaults to :obj:`True`):
+            If :obj:`True`, use randomly initialized paraeters for the Question-Aware Span Selection head on top of
+            Splinter.
+        question_token_id (:obj:`int`, `optional`, defaults to :obj:`104`):
```
```suggestion
        question_token_id (:obj:`int`, `optional`, defaults to 104):
```

patrickvonplaten commented on src/transformers/models/splinter/modeling_splinter.py:
```diff
@@ -0,0 +1,1556 @@
+            cross_attentions=all_cross_attentions,
+        )
+
+
```
# Copied from transformers.models.bert.modeling_bert.BertPredictionHeadTransform with Bert->Splinter

oriram replied:
We will remove these models any way :)

patrickvonplaten commented on src/transformers/models/splinter/modeling_splinter.py:
```diff
@@ -0,0 +1,1556 @@
+        return hidden_states
+
+
+class SplinterLMPredictionHead(nn.Module):
```
```# Copied from transformers.models.bert.modeling_bert.BertLMPredictionHead with Bert->Splinter```

patrickvonplaten commented on src/transformers/models/splinter/modeling_splinter.py:
```diff
@@ -0,0 +1,1556 @@
+        return hidden_states
+
+
+class SplinterOnlyMLMHead(nn.Module):
```
`# Copied from transformers.models.bert.modeling_bert.BertOnlyMLMHead with Bert->Splinter`

patrickvonplaten commented on src/transformers/models/splinter/modeling_splinter.py:
```diff
@@ -0,0 +1,1556 @@
+        return reordered_past
+
+
+class SplinterClassificationHead(nn.Module):
```
Here we can also add a #Copied from no?

oriram replied:
Removed this class :)

patrickvonplaten commented on src/transformers/models/splinter/modeling_splinter.py:
```diff
@@ -0,0 +1,1556 @@
+    return gathered_output
+
+
+class FullyConnectedLayer(nn.Module):
```
```suggestion
class SplinterFullyConnectedLayer(nn.Module):
```

patrickvonplaten commented on src/transformers/models/splinter/modeling_splinter.py:
```diff
@@ -0,0 +1,1556 @@
+        self.act_fn = ACT2FN[hidden_act]
+        self.LayerNorm = nn.LayerNorm(self.output_dim)
+
+    def forward(self, inputs):
```
can we replace all `temp` with `hidden_states` for consistency?

oriram replied:
Done :)

patrickvonplaten commented on src/transformers/models/splinter/modeling_splinter.py:
```diff
@@ -0,0 +1,1556 @@
+
+class FullyConnectedLayer(nn.Module):
+    def __init__(self, input_dim, output_dim, hidden_act="gelu"):
+        super(FullyConnectedLayer, self).__init__()
```
```suggestion
        super().__init__()
```

patrickvonplaten commented on src/transformers/models/splinter/modeling_splinter.py:
```diff
@@ -0,0 +1,1556 @@
+        start_reps = self.start_transform(inputs)  # [batch_size, seq_length, dim]
+        end_reps = self.end_transform(inputs)  # [batch_size, seq_length, dim]
+
+        temp = self.start_classifier(query_start_reps)  # [batch_size, num_positions, dim]
```
replace `temp` with `hidden_states`

oriram replied:
Done :)

patrickvonplaten commented on src/transformers/models/splinter/modeling_splinter.py:
```diff
@@ -0,0 +1,1556 @@
+        return start_logits, end_logits
+
+
+class ClassificationHead(nn.Module):
```
```suggestion
class SplinterClassificationHead(nn.Module):
```

oriram replied:
Actually this class is outdated so I removed it :)

patrickvonplaten commented on src/transformers/models/splinter/modeling_splinter.py:
```diff
@@ -0,0 +1,1556 @@
+                )
+            else:
+                question_position_for_each_example = torch.zeros(
+                    inputs_embeds.size(0), dtype=torch.int64, layout=inputs_embeds.layout, device=inputs_embeds.device
```
```suggestion
                    inputs_embeds.size(0), dtype=torch.long, layout=inputs_embeds.layout, device=inputs_embeds.device
```

patrickvonplaten commented on src/transformers/models/splinter/modeling_splinter.py:
```diff
@@ -0,0 +1,1556 @@
+            question_positions_were_none = True
+
+        if attention_mask is not None:
+            attention_mask[input_ids == 102] = 0
```
This shouldn't be hardcoded here I think - what token is 102?

oriram replied:
Removed this part, it's not crucial

patrickvonplaten commented on src/transformers/models/splinter/modeling_splinter.py:
```diff
@@ -0,0 +1,1556 @@
+
+        self.splinter = SplinterModel(config)
+        self.initialize_new_qass = config.initialize_new_qass
+        self.splinter_qass = QuestionAwareSpanSelectionHead(config) if not self.initialize_new_qass else None
```
I don't really understand why we have `self.splinter_qass` which is the exact same module as `self.new_splinter_qass` and it's always only one of the 2 -> why not just have one self.splinter_qass`?

patrickvonplaten replied:
Is it because of the naming in the weights?

oriram replied:
Exactly. This was our way to differentiate between keeping the weights of this layer and initializing them randomly. 
Any suggestions will be welcome, it's pretty ugly :)

oriram replied:
By the way, is there any way we can do it while keeping the logging (which of the weights from the checkpoint are not used and vice versa) correct?

patrickvonplaten replied:
Yeah in this case, we should have multiple checkpoints I think.
One checkpoint which **does have** the trained weights for `spliter_qass` and one checkpoint which does not have them. For the checkpoint that does not have the weights for `spliter_qass` they will automatically be randomely initialized

oriram replied:
Great.. So I think the way to control it will be via `--model_name_or_path`, right?
In other words, I'll remove it from `SplinterConfig`.
Unless you had other thoughts?

patrickvonplaten commented on src/transformers/models/splinter/tokenization_splinter.py:
```diff
@@ -0,0 +1,164 @@
+}
+
+
+class SplinterTokenizer(BertTokenizer):
```
Let's not abstract from a different tokenizer here. It's better to just copy-paste the code and make use of `#Copied from`

oriram replied:
@patrickvonplaten We do have a bit of a different tokenization (mainly in terms of `build_inputs_with_special_tokens`)

patrickvonplaten replied:
Sorry, I wasn't super clear here I think. We are trying to remove as many abstractions as possible which in this case would mean that we don't abstract from the `BertTokentizer` class. Instead it would be better to just have: 

```python
class SplinterTokenizer:
```

and just copy paste all the required functionality from the BertTokenizer

patrickvonplaten commented on tests/test_modeling_splinter.py:
```diff
@@ -0,0 +1,474 @@
+
+
+@require_torch
+class SplinterModelIntegrationTest(unittest.TestCase):
```
Can we add one hard-coded QA integration test making sure that the important QA models work correctly?

patrickvonplaten replied:
Awesome :-)

patrickvonplaten commented:
Thanks a lot for the PR! Think we can merge this soon :-) Some points that I think will be important to adapt before merging are:

- Simplify the logic of `splinter_qass` vs `new_splinter_qass`. IMO there should only be one `splinter_class` class attribute and if this has to be reinitialized or set to 0 we could instead add a `reinit` function. I don't really understand why we need to identical `splinter_qass` and `new_spliter_qass` modules
- Make sure we don't have hardcoded id's such as 102 in the model
- Add a QA integration test to make sure the model works as expected

oriram committed:
Some minor quality changes

oriram committed:
Update docs/source/model_doc/splinter.rst

Co-authored-by: Patrick von Platen <patrick.v.platen@gmail.com>

oriram committed:
Update src/transformers/models/splinter/configuration_splinter.py

Co-authored-by: Patrick von Platen <patrick.v.platen@gmail.com>

oriram committed:
Change FullyConnectedLayer to SplinterFullyConnectedLayer

oriram committed:
Variable naming

oriram committed:
Reove gather_positions function

oriram commented on src/transformers/models/splinter/modeling_splinter.py:
```diff
@@ -0,0 +1,1556 @@
+        self.end_classifier = nn.Linear(config.hidden_size, config.hidden_size, bias=False)
+
+    def forward(self, inputs, positions):
+        gathered_reps = gather_positions(inputs, positions)
```
Done :)

oriram committed:
Remove ClassificationHead as it's outdated

oriram committed:
Update src/transformers/models/splinter/modeling_splinter.py

Co-authored-by: Patrick von Platen <patrick.v.platen@gmail.com>

oriram committed:
Remove hardcoded 102 token id

oriram committed:
Merge branch 'add_splinter' of https://github.com/yuvalkirstain/transformers into add_splinter

oriram committed:
Minor style change

oriram committed:
Added "tau" organization to all model identifiers & URLS

oriram committed:
Added tau to the tests as well

oriram commented:
@patil-suraj @patrickvonplaten @sgugger 
Any idea why this exception is raised when calling `make quality`?
```
Traceback (most recent call last):
  File "/mnt/c/Program Files/JetBrains/PyCharm 2019.3.2/plugins/python/helpers/pydev/pydevd.py", line 1434, in _exec
    pydev_imports.execfile(file, globals, locals)  # execute the script
  File "/mnt/c/Program Files/JetBrains/PyCharm 2019.3.2/plugins/python/helpers/pydev/_pydev_imps/_pydev_execfile.py", line 18, in execfile
    exec(compile(contents+"\n", file, 'exec'), glob, loc)
  File "/mnt/c/Users/ori/Desktop/Ori/Research/transformers_splinter/utils/check_copies.py", line 353, in <module>
    check_copies(args.fix_and_overwrite)
  File "/mnt/c/Users/ori/Desktop/Ori/Research/transformers_splinter/utils/check_copies.py", line 186, in check_copies
    new_diffs = is_copy_consistent(filename, overwrite)
  File "/mnt/c/Users/ori/Desktop/Ori/Research/transformers_splinter/utils/check_copies.py", line 164, in is_copy_consistent
    theoretical_code = blackify(lines[start_index - 1] + theoretical_code)
  File "/mnt/c/Users/ori/Desktop/Ori/Research/transformers_splinter/utils/check_copies.py", line 104, in blackify
    result = black.format_str(code, mode=black.FileMode([black.TargetVersion.PY35], line_length=119))
  File "/home/oriram/venv/transformers_splinter/lib/python3.7/site-packages/black/__init__.py", line 1063, in format_str
    src_node = lib2to3_parse(src_contents.lstrip(), mode.target_versions)
  File "/home/oriram/venv/transformers_splinter/lib/python3.7/site-packages/black/__init__.py", line 1171, in lib2to3_parse
    raise exc from None
black.InvalidInput: Cannot parse: 15:4:     def __init__(self, config, add_pooling_layer=True):
```
Didn't happen before. Tried to debug it, but couldn't understand the cause.
Also, it seems like it's not in any file related to Splinter, as I don't have the argument `add_pooling layer`.

sgugger commented on src/transformers/models/splinter/modeling_splinter.py:
```diff
@@ -0,0 +1,1531 @@
+"""
+
+
+# Copied from transformers.models.bert.modeling_bert.BertModel with Bert->Splinter
```
This line needs to be directly on top of the class, this is probably the cause of the error you are seeing.

oriram committed:
Copy-from comments

oriram committed:
Removed all unnecessary classes (e.g. SplinterForMaskedLM)

oriram committed:
Running make fix-copies

oriram committed:
Bug fix: Further removed unnecessary classes

oriram committed:
Merge branch 'master' into add_splinter

oriram committed:
Add Splinter to AutoTokenization

oriram commented:
@sgugger 
One idea I had in mind regarding the `self.splinter_qass` and `self.new_splinter_qass` was just to create two more checkpoints (so overall there will be 4 rather than 2): 
```
tau/splinter-base-with-qass 
tau/splinter-base
tau/splinter-large-with-qass 
tau/splinter-large
```
In the ones without qass, I'll drop the weights from `pytorch_model.bin`. 
This will also keep the message `All the weights of SplinterForQuestionAnswering were initialized from the model checkpoint at X` correct.
Does that sound OK?

sgugger commented:
I imagine there will be an associated config parameter to determine which layer use?
It looks like a good idea.

sgugger commented:
As a result of #13023 , you will need to rebase your PR on master and solve the merge conflicts (basically, you will just need to re-add the models/tokenizers and config in the auto-mappings as strings). Let us know if you need any help with that.

oriram committed:
Add an integration test for Splinter

oriram committed:
Removed initialize_new_qass from config - It will be done through different checkpoints

oriram committed:
Removed `initialize_new_qass` from documentation as well

oriram committed:
Added new checkpoint names (`tau/splinter-base-qass` and same for large) in the code

oriram committed:
Minor change to test

oriram committed:
SplinterTokenizer now doesn't abstract from BertTokenizer

oriram committed:
SplinterTokenizerFast also dosn't abstract from Bert

oriram committed:
style and quality

oriram commented:
Hi @sgugger @patil-suraj @patrickvonplaten,
Sorry to bother you, but I couldn't find the reason as to why the `run_tests_tf` (and the two others) fail.
CircleCI doesn't provide much info..
Other than that and the rebase, I think everything is set to do the PR, took care of:

- Removing abstractions from Bert classes
- Separating the models into `splinter-base` and `splinter-base-qass` (and same for `large`), as well as removing all references to `vonfig.initialize_new_qass` from the code
- Added an integration test
- Removed all unnecessary classes (`SplinterForMaskedLM` etc.)
- Fixed `Copy from` issues

Thanks!!

sgugger commented on tests/test_modeling_splinter.py:
```diff
@@ -0,0 +1,219 @@
+
+import unittest
+
+import torch
```
This import should be moved in the `is_torch_available`, that's the reason the TF and Flax tests fail.

oriram committed:
bug fix: import ing torch in tests only if it's available

oriram commented:
Thanks @sgugger for your quick response!
I think the tests are now failing due to the rebase issue, as the errors don't seem related to Splinter.
Any chance you can help with the rebase? Don't have any experience with that..
Would really appreciate it :)

LysandreJik committed:
Auto mappings

LysandreJik committed:
Merge branch 'master' into add_splinter_2

LysandreJik commented:
Hello @oriram, I just took care of the merge and the auto-classes

oriram commented:
Many thanks @LysandreJik @sgugger!!
Are we ready to merge then?

patrickvonplaten commented on src/transformers/models/splinter/modeling_splinter.py:
```diff
@@ -0,0 +1,932 @@
+            start_positions.clamp_(0, ignored_index)
+            end_positions.clamp_(0, ignored_index)
+
+            loss_fct = CrossEntropyLoss(ignore_index=ignored_index)
```
Usually we don't overwrite the default `-100` with a unique `ignored_index`, but I think it's fine in this case as there is no padding on the targets that would have to be replaced with `-100`

LysandreJik replied:
Ok, I'm not opposed to that

patrickvonplaten reviewed (approved):
Did a final pass and looks good to me - think we can merge :-)

LysandreJik commented on src/transformers/models/splinter/configuration_splinter.py:
```diff
@@ -0,0 +1,128 @@
+
+        Example::
+
+        >>> from transformers import SplinterModel, SplinterConfig
+
+        >>> # Initializing a Splinter tau/splinter-base style configuration
+        >>> configuration = SplinterConfig()
+
+        >>> # Initializing a model from the tau/splinter-base style configuration
+        >>> model = SplinterModel(configuration)
+
+        >>> # Accessing the model configuration
+        >>> configuration = model.config
```
Please add an indentation level

oriram replied:
Hi @LysandreJik, what do you mean?

LysandreJik replied:
Let me offer a suggestion :)

LysandreJik commented on src/transformers/models/splinter/modeling_splinter.py:
```diff
@@ -0,0 +1,932 @@
+# coding=utf-8
+# Copyright 2021 Ori Ram, Yuval Kirstain, Jonathan Berant, Amir Globerson, Omer Levy The HuggingFace Inc. team. All rights reserved.
```
Same here regarding copyright

LysandreJik commented on src/transformers/models/splinter/modeling_splinter.py:
```diff
@@ -0,0 +1,1629 @@
+    return x * torch.tanh(nn.functional.softplus(x))
+
+
+class SplinterEmbeddings(nn.Module):
```
I think this was resolved without being taken care of; the code will be updated to register the `token_type_ids` non-persistent buffer, but I don't think that's an issue? (it is necessary for torchscript scripting across different devices)

LysandreJik commented on src/transformers/models/splinter/tokenization_splinter.py:
```diff
@@ -0,0 +1,531 @@
+# coding=utf-8
+# Copyright Ori Ram, Yuval Kirstain, Jonathan Berant, Amir Globerson, Omer Levy and The HuggingFace Inc. team.
```
Same here regarding copyright

LysandreJik commented on src/transformers/models/splinter/tokenization_splinter.py:
```diff
@@ -0,0 +1,531 @@
+    return tokens
+
+
+class SplinterTokenizer(PreTrainedTokenizer):
```
Is this tokenizer any different from BERT or is it the exact same?

oriram replied:
Hi @LysandreJik,
Same as above - The tokenizer is indeed different (see `build_inputs_with_special_tokens` )

LysandreJik reviewed (approved):
This looks great, thanks for your contribution @oriram, @yuvalkirstain!

My only question is relative to the BERT tokenizer - I wonder if it's the exact same (in which case the `BERT` tokenizer should be defined as default tokenizer and this tokenizer removed) or if it is different, in which case it should be kept.

oriram commented on src/transformers/models/splinter/configuration_splinter.py:
```diff
@@ -0,0 +1,128 @@
+# coding=utf-8
+# Copyright Ori Ram, Yuval Kirstain, Jonathan Berant, Amir Globerson, Omer Levy and The HuggingFace Inc. team. All rights reserved.
```
Hi @LysandreJik , 
the tokenizer is a bit different because of the special `[QUESTION]` token - See the 'build_inputs_with_special_tokens' function

oriram committed:
Changed copyrights in Splinter's files

oriram commented:
Hi @LysandreJik :)

- Changed copyrights in Splinter's 4 files
- Regarding your comments on `SplinterTokenizer` - The difference stems from dealing with the special `[QUESTION]` token which is used for building question representations

Many thanks!

LysandreJik commented on src/transformers/models/splinter/configuration_splinter.py:
```diff
@@ -0,0 +1,128 @@
+        question_token_id (:obj:`int`, `optional`, defaults to 104):
+            The id of the ``[QUESTION]`` token.
+
+        Example::
+
+        >>> from transformers import SplinterModel, SplinterConfig
+
+        >>> # Initializing a Splinter tau/splinter-base style configuration
+        >>> configuration = SplinterConfig()
+
+        >>> # Initializing a model from the tau/splinter-base style configuration
+        >>> model = SplinterModel(configuration)
+
+        >>> # Accessing the model configuration
+        >>> configuration = model.config
```
```suggestion
        Example::

            >>> from transformers import SplinterModel, SplinterConfig

            >>> # Initializing a Splinter tau/splinter-base style configuration
            >>> configuration = SplinterConfig()

            >>> # Initializing a model from the tau/splinter-base style configuration
            >>> model = SplinterModel(configuration)

            >>> # Accessing the model configuration
            >>> configuration = model.config
```

LysandreJik reviewed (approved):
LGTM, thanks for this great work!

oriram committed:
Update src/transformers/models/splinter/configuration_splinter.py

Co-authored-by: Lysandre Debut <lysandre@huggingface.co>

oriram commented:
Great @LysandreJik!
Let's merge? :)

oriram commented:
@patil-suraj @sgugger @LysandreJik @patrickvonplaten 
Just wanted to say many thanks again for all your effort in this PR!!

patrickvonplaten commented:
@oriram - thanks a mille for your great PR! Let's try to promote Splinter as much as possible :-)
